design_documents/transactional_testing/transactional_testing.md

```diff
@@ -0,0 +1,172 @@
+2. The test scripts are compiled with foundry (until a plugin solution is available)
+3. The test uses artifacts from Ad. 1, runs program from Ad 2. on own VM with hint processing logic 
+
+### File organization within scarb workspace
+
+Each package can have its own contract, so each package could have its own e2e directory, 
+and the whole workspace could have it as well (to test it in integration).
+For example:
+```
+./project
+  ./Scarb.toml
+  ./e2e
+    ./e2e/test_integration.cairo
+    ./e2e/test_integration2.cairo
+  ./pkg1
+    ./src
+    ./test
+    ./e2e
+      ./test_contract_1.cairo
+  ./pkg2
+  ...
+```
+## Solution analysis
+
+### Pros
+
+- Support of external tech (devnet, katana) via specialized libraries
+- Flow close to real starknet transactions, fees, testing how the protocol changes affect contracts
+- Ability to test accounts (__validate__, __execute__)
+- Ability to test real deployment/interaction scripts in environment close to real one
+
+### Cons
+- Keeping up the specialized libraries updated as tech adds new features/changes
+- Overall higher maintenance cost
+- Test execution speed significantly lower (depending on the used endpoints' throughput)
+- Dry-run mode from cast scripts might be sufficient for some use cases (testing deployments)
+- Might be confusing to users if not communicated correctly
+
+
+## Usage examples
+1. Test transaction signing
+```
+from xxx_std import {deploy_account, deploy_contracts, end_txn, call, stark_curve, Transaction, TransactionStatus, Calldata, Call};  
+from xxx_devnet_extras import {set_time, dump};
+
+const PREFUNDED_ADDRESS = 0x31231; // Set by the user, or generated inside the test
+
+#[devnet_test(url=localhost:3000)]
+fn test_signature_validation() {
+    let account_contract = deploy_account("OZAccount", PREFUNDED_ADDRESS, );
+    let mock_contract = deploy_contract("MockContract"); // The one that account is calling
+
+    let txn = Transaction(
+        version=2,      // Different versions could be supported
+        address=account_contract.address,
+        function="__execute__", 
+        calldata=Calldata::from_calls(array![Call(...), Call(...)])),
+        max_fee=100000,
+    );
+    
+    let signed_txn = stark_curve::sign(txn); // This would also be pluggable, for users to be able to sign with different curves
+    
+    set_time(123); // Set time for txn
+    let result = send_txn(signed_txn); // Synchronous
+    match result {
+        TransactionStatus::Rejected(data) => { /* assert failure data or gas */ },
+        TransactionStatus::Accepted(data) => { /* assert result data  or gas */ }
+    };
+    
+    let account_balance = call(
+        address=mock_contract.address, 
+        function="get_balance",
+        calldata=Calldata::from_calls(array![Call(...), Call(...)])
+    );
+    
+    assert(account_balance > ..., "not correct balance");
+    dump(); // Dump can be used for pre-populating your test environment
+}
+```
```

Comment:
Looking at this example usage I wonder if it actually would make sense to just follow the alternative approach 1. and extend the deployment scripts.

I see that syntax is a bit different, but it seems to me like the backed for this will be very, very similar to cast: Not only the runtime part, but also just logic for making rpc calls to the whatever devnet we're using.